src/ray/util/shared_lru.h

```diff
@@ -0,0 +1,203 @@
+    lru_list_.erase(cache_iter->second.lru_iterator);
+    cache_.erase(cache_iter);
+
+    // Re-insert into the cache, no need to check capacity.
+    lru_list_.emplace_front(key);
+    Entry new_entry{value, lru_list_.begin()};
+    cache_[key] = std::move(new_entry);
+
+    return value;
+  }
+
+  // Clear the cache.
+  void Clear() {
+    cache_.clear();
+    lru_list_.clear();
+  }
+
+  // Accessors for cache parameters.
+  size_t max_entries() const { return max_entries_; }
+
+ private:
+  struct Entry {
+    // The entry's value.
+    Val value;
+
+    // A list iterator pointing to the entry's position in the LRU list.
+    typename std::list<Key>::iterator lru_iterator;
+  };
+
+  using EntryMap = absl::flat_hash_map<Key, Entry>;
+
+  // The maximum number of entries in the cache. A value of 0 means there is no
+  // limit on entry count.
+  const size_t max_entries_;
+
+  // All keys are stored as refernce (`std::reference_wrapper`), and the
+  // ownership lies in `lru_list_`.
+  EntryMap cache_;
+
+  // The LRU list of entries. The front of the list identifies the most
+  // recently accessed entry.
+  std::list<Key> lru_list_;
+};
+
+// Same interfaces as `SharedLruCache`, but all cached values are
+// `const`-specified to avoid concurrent updates.
+template <typename K, typename V>
+using SharedLruConstCache = SharedLruCache<K, const V>;
+
+// Same interface and functionality as `SharedLruCache`, but thread-safe version.
+template <typename Key, typename Val>
+class ThreadSafeSharedLruCache final {
+ public:
+  using key_type = Key;
+  using mapped_type = Val;
+
+  // A `max_entries` of 0 means that there is no limit on the number of entries
+  // in the cache.
+  explicit ThreadSafeSharedLruCache(size_t max_entries) : cache_(max_entries) {}
+
+  ThreadSafeSharedLruCache(const ThreadSafeSharedLruCache &) = delete;
+  ThreadSafeSharedLruCache &operator=(const ThreadSafeSharedLruCache &) = delete;
+
+  ~ThreadSafeSharedLruCache() = default;
+
+  // Insert `value` with key `key`. This will replace any previous entry with
+  // the same key.
+  void Put(Key key, Val value) {
+    std::lock_guard lck(mtx_);
+    cache_.Put(std::move(key), std::move(value));
+  }
+
+  // Delete the entry with key `key`. Return true if the entry was found for
+  // `key`, false if the entry was not found. In both cases, there is no entry
+  // with key `key` existed after the call.
+  bool Delete(const Key &key) {
+    std::lock_guard lck(mtx_);
+    return cache_.Delete(key);
+  }
+
+  // Look up the entry with key `key`. Return std::nullopt if key doesn't exist.
+  // If found, return a copy for the value.
+  std::optional<Val> Get(const Key &key) {
+    std::lock_guard lck(mtx_);
+    return cache_.Get(key);
+  }
+
+  // Clear the cache.
+  void Clear() {
+    std::lock_guard lck(mtx_);
+    cache_.Clear();
+  }
+
+  // Accessors for cache parameters.
+  size_t max_entries() const { return cache_.max_entries(); }
+
+ private:
+  std::mutex mtx_;
```

Comment:
I feel we don't omit vowels in variable names. Either `m_` or `mutex_`.

Reply:
That's too picky...
I personally don't like either of you suggestions.
- I would like to avoid short names as possible, for example, `es` (execution state), except `i` and `j` which everybody knows;
- `mutex` is already a type name, having variable (mostly) conflict with type name hurts readability and code seachability
- quick search: there're syscalls using `mtx` to indicate mutex: https://manpages.ubuntu.com/manpages/trusty/man9/mtx_lock.9freebsd.html

Reply:
I actually have no preference among the names but I'd like to be consistent within the codebase. Currently in the codebase it's either `mu_` or `mutex_` (a bit unfortunate) so let's not introduce another name (unless you want to just renaming all the existing ones to `mtx_` which is fine to me)?

Reply:
OK, updated to `mu_`.